### ISSUE TITLE
Negation and None

### ISSUE THREAD

**Issue by socketpair** [PyCQA/pylint#5085]
### Bug description

```python
Pylint incorrectly tracks types:


class Xxx:
    def __init__(self):
        self.__val = None

    @property
    def get_value(self) -> int:
        if self.__val is None:
            self.__val = 42
        return self.__val

    def method(self):
        return 'qwe'[:-self.get_value]
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
a.py:12:22: E1130: bad operand type for unary -: NoneType (invalid-unary-operand-type)
```

### Expected behavior

no errors

### Pylint version

```shell
pylint 2.9.6
astroid 2.6.6
Python 3.9.7 (default, Aug 30 2021, 00:00:00) 
[GCC 11.2.1 20210728 (Red Hat 11.2.1-1)]
```

### OS / Environment

Fedora 34

**Comment by saroad2** [comment]
I also encountered this bug with the following example:

```python
from pathlib import Path
from typing import Optional, List


def clear(paths_list: List[Path], limit: Optional[int] = None):
    if limit is not None and limit < len(paths_list):
        paths_list = paths_list[-limit:]
    for path in paths_list:
        path.unlink()
```

When running *Pylint* over the code above it raises:
```
my\path.py:7:32: E1130: bad operand type for unary -: NoneType (invalid-unary-operand-type)
```

**Comment by sam-s** [comment]
Here is an even smaller test case:
```
l = [1,2,3]
x = None
if isinstance(x, int):
    l = l[-x:]
```
that gives
```
 bad operand type for unary -: NoneType (invalid-unary-operand-type)
```
on
```
pylint 2.17.5
astroid 2.15.6
Python 3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)]
```

**Comment by socketpair** [comment]
```python
val: int | None = None


def get_value() -> int:
    if val is None:
        return 42
    return val


print(None + 1) # NO ERROR HERE ?! WUT ?
print(get_value() + 1)  # NO ERROR HERE!!!
print(-get_value())   # BUMM! error! WHY ?
```

```
pylint 3.0.4
astroid 3.0.3
Python 3.12.2 (main, Feb 21 2024, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)]
```

**Comment by socketpair** [comment]
@Pierre-Sassoulas @sam-s @saroad2 

Do you have any ideas ?

**Comment by Pierre-Sassoulas** [comment]
The ``is None`` case was introduced in astroid 2.13 https://github.com/pylint-dev/astroid/pull/1189, ``isinstance`` still needs works in astroid. I won't have enough time to dig into the specifics about how the inference works in your specific case @socketpair .

**Comment by socketpair** [comment]
@Pierre-Sassoulas I has minified:

```python
val = None
if val is None:
    val = 42
print(-val)
```

`E1130: bad operand type for unary -: NoneType (invalid-unary-operand-type)`

Still requires digging ?
